apps/app_gps.c

```diff
@@ -693,316 +1001,446 @@ static void *gpsthread(void *data)
+	} else {
 		deflon = NULL;
+	}
 	val = (char *) ast_variable_retrieve(cfg, ctg, "elev");
+	if (val) {
+		defelev = ast_strdupa(val);
+	} else {
+		defelev = NULL;
+	}
 	val = (char *) ast_variable_retrieve(cfg, ctg, "interval");
-	if (val)
+	if (val) {
 		interval = atoi(val);
-	else
+	} else {
 		interval = GPS_UPDATE_SECS;
+	}
 	val = (char *) ast_variable_retrieve(cfg, ctg, "ehlert");
-	if (val)
+	if (val) {
 		ehlert = ast_true(val);
-	else
+	} else {
 		ehlert = 0;
+	}
 	ast_config_destroy(cfg);
 	cfg = NULL;
-	time(&lastupdate);
+	/*
+	 * Set the default position for this section
+	 * If it is [general], we already have the defaults
+	 * otherwise, we need to build the specific defaults
+	 * for this section.
+	 */
+	 if (!strcmp(ctg, "general") || (!deflat && !deflon)) {
+		 this_def_position = general_def_position;
+	 } else {
+		this_def_position.is_valid = 1;
+		lat_decimal_to_DMS(strtof(deflat, NULL), this_def_position.latitude, sizeof(this_def_position.latitude));
+		lon_decimal_to_DMS(strtof(deflon, NULL), this_def_position.longitude, sizeof(this_def_position.longitude));
+		/* See if we have a default elevation */
+		if (defelev) {
+			float eleva, elevd;
+			eleva = strtof(defelev, NULL);
+			elevd = (eleva - floor(eleva)) * 10 + 0.5;
+			snprintf(this_def_position.elevation, sizeof(this_def_position.elevation), "%03d.%1d", (int) eleva, (int) elevd);
+		} else {
+			strcpy(this_def_position.elevation, "000.0");
+		}
+	 }
+	
+	memset(&selected_position, 0, sizeof(selected_position));
+	lastupdate = time_monotonic();
 	my_update_secs = GPS_UPDATE_SECS;
+	
 	while (run_forever) {
-		gotfiledata = 0;
-		if (!strcmp(ctg, "general"))
-			strcpy(fname, GPS_DATA_FILE);
-		else
-			snprintf(fname, sizeof(fname) - 1, GPS_SUB_FILE, ctg);
-		time(&now);
-		fp = fopen(fname, "r");
-		if (fp && (fstat(fileno(fp), &mystat) != -1) && (mystat.st_size < 100)) {
-			if (fscanf(fp, "%u %s %s", &u, lat, lon) == 3) {
-				was = (time_t) u;
-				if ((was + GPS_VALID_SECS) >= now) {
-					gotfiledata = 1;
-					if (now >= (lastupdate + my_update_secs)) {
-						report_aprs(ctg, lat, lon);
-						lastupdate = now;
-						my_update_secs = interval;
-					}
-				}
+		now = time_monotonic();
+		
+		ast_mutex_lock(&position_update_lock);
+		selected_position.is_valid = 0;
+		/* See if we need to send live GPS or the default */
+		if (current_gps_position.is_valid) {
+			selected_position = current_gps_position;
+		} else {
+			if (this_def_position.is_valid && !ehlert) {
+				selected_position = this_def_position;
+				selected_position.last_updated = time(NULL);
 			}
 		}
-		if ((!gotfiledata) && (!ehlert) && deflat && deflon) {
+		ast_mutex_unlock(&position_update_lock);
+		/* 
+		 * See if it is time to send the position report
+		 * The last_updated time must be current so that
+		 * we know we are getting good GPS information.
+		 */
+		if (selected_position.is_valid && (selected_position.last_updated + GPS_VALID_SECS) >= time(NULL)) {
```

Comment:
I don't know - maybe all in one if statement?   I don't see a reason for the second `if`